src/hotspot/share/opto/vectorIntrinsics.cpp

```diff
@@ -2850,3 +2850,146 @@ bool LibraryCallKit::inline_vector_compress_expand() {
+  }
+
+  if (!is_klass_initialized(vector_klass)) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** klass argument not initialized");
+    }
+    return false;
+  }
+
+  ciType* elem_type = elem_klass->const_oop()->as_instance()->java_mirror_type();
+  if (!elem_type->is_primitive_type()) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** not a primitive bt=%d", elem_type->basic_type());
+    }
+    return false; // should be primitive type
+  }
+
+  int num_elem = vlen->get_con();
+  BasicType elem_bt = elem_type->basic_type();
+  BasicType iota_bt = elem_bt;
+  // The iota indices are integral type
+  if (is_floating_point_type(elem_bt)) {
+    iota_bt = elem_bt == T_FLOAT ? T_INT : T_LONG;
+  }
+
+  // Get the vector add/mul op based on the basic element type
+  int add_op = VectorSupport::vop2ideal(VectorSupport::VECTOR_OP_ADD, elem_bt);
+  int mul_op = VectorSupport::vop2ideal(VectorSupport::VECTOR_OP_MUL, elem_bt);
+  int vadd_op = VectorNode::opcode(add_op, elem_bt);
+  int vmul_op = VectorNode::opcode(mul_op, elem_bt);
+
+  // Check whether the basic ops are supported by the current hardware
+  if (!arch_supports_vector(Op_VectorLoadConst, num_elem, iota_bt, VecMaskNotUsed) ||
+      !arch_supports_vector(vmul_op, num_elem, elem_bt, VecMaskNotUsed) ||
+      !arch_supports_vector(vadd_op, num_elem, elem_bt, VecMaskNotUsed)) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** not supported: vlen=%d etype=%s", num_elem, type2name(elem_bt));
+    }
+    return false; // not supported
+  }
+
+  // If it is a floating point type vector, additionally check whether the
+  // relative vector cast op is supported by the current hardware
+  if (is_floating_point_type(elem_bt)) {
+    int vcast_op = elem_bt == T_FLOAT ? Op_VectorCastI2X : Op_VectorCastL2X;
+    if (!arch_supports_vector(vcast_op, num_elem, elem_bt, VecMaskNotUsed)) {
+      if (C->print_intrinsics()) {
+        tty->print_cr("  ** not supported: vlen=%d etype=%s", num_elem, type2name(elem_bt));
+      }
+      return false; // not supported
+    }
+  }
+
+  ciKlass* vbox_klass = vector_klass->const_oop()->as_instance()->java_lang_Class_klass();
+  const TypeInstPtr* vbox_type = TypeInstPtr::make_exact(TypePtr::NotNull, vbox_klass);
+  Node* opd = unbox_vector(argument(3), vbox_type, elem_bt, num_elem);
+  if (opd == NULL) {
+    if (C->print_intrinsics()) {
+      tty->print_cr("  ** unbox failed vector=%s",
+                    NodeClassNames[argument(3)->Opcode()]);
+    }
+    return false;
+  }
+
+  // Compute the iota indice vector
+  const TypeVect* vt = TypeVect::make(elem_bt, num_elem);
+  const TypeVect* iota_vt = TypeVect::make(iota_bt, num_elem);
+  Node* iota = gvn().transform(new VectorLoadConstNode(gvn().makecon(TypeInt::ZERO), iota_vt));
+  if (elem_bt == T_FLOAT) {
+    iota = gvn().transform(new VectorCastI2XNode(iota, vt));
+  } else if (elem_bt == T_DOUBLE) {
+    iota = gvn().transform(new VectorCastL2XNode(iota, vt));
+  }
```

Comment:
Since we are loading constants from stub initialized memory locations, defining new stubs for floating point iota indices may eliminate need for costly conversion instructions. Specially on X86 conversion b/w Long and Double is only supported by AVX512DQ targets and intrinsification may fail for legacy targets.

Reply:
Make sense to me! I'v changed the codes based on the suggestion in the latest commit. Please help to take a review again! Thanks a lot!